arbos/programs/programs.go

```diff
@@ -308,30 +308,44 @@ func getWasm(statedb vm.StateDB, program common.Address) ([]byte, error) {
 }
 
 func (p Programs) getProgram(contract *vm.Contract) (Program, error) {
-
 	return p.deserializeProgram(contract.CodeHash)
 }
 
 func (p Programs) deserializeProgram(codeHash common.Hash) (Program, error) {
 	data, err := p.programs.Get(codeHash)
+	if err != nil {
+		return Program{}, err
+	}
+	compiledHash, err := p.moduleHashes.Get(codeHash)
+	if err != nil {
+		return Program{}, err
+	}
 	return Program{
-		wasmSize:  arbmath.BytesToUint16(data[26:28]),
-		footprint: arbmath.BytesToUint16(data[28:30]),
-		version:   arbmath.BytesToUint16(data[30:]),
-	}, err
+		wasmSize:   arbmath.BytesToUint16(data[26:28]),
+		footprint:  arbmath.BytesToUint16(data[28:30]),
+		version:    arbmath.BytesToUint16(data[30:]),
+		moduleHash: compiledHash,
+	}, nil
 }
 
-func (p Program) serialize() common.Hash {
+func (p Programs) setProgram(codehash common.Hash, program Program) error {
 	data := common.Hash{}
-	copy(data[26:], arbmath.Uint16ToBytes(p.wasmSize))
-	copy(data[28:], arbmath.Uint16ToBytes(p.footprint))
-	copy(data[30:], arbmath.Uint16ToBytes(p.version))
-	return data
+	copy(data[26:], arbmath.Uint16ToBytes(program.wasmSize))
+	copy(data[28:], arbmath.Uint16ToBytes(program.footprint))
+	copy(data[30:], arbmath.Uint16ToBytes(program.version))
+	err := p.programs.Set(codehash, data)
+	if err != nil {
+		return err
+	}
+	return p.moduleHashes.Set(codehash, program.moduleHash)
 }
 
 func (p Programs) CodehashVersion(codeHash common.Hash) (uint16, error) {
 	program, err := p.deserializeProgram(codeHash)
```

Comment:
Maybe this is fine, but this also loads the module hash when it doesn't need to

Reply:
You only ever call this during activation. Perhaps we could change the name?

Reply:
Oh, I was looking at the wrong line -- def agree, I've actually gone ahead and removed `program.moduleHash`